src/openzeppelin/upgrades/upgradeable.cairo

```diff
@@ -0,0 +1,58 @@
+use starknet::class_hash::ClassHash;
+
+trait UpgradeableTrait<TState> {
+    fn upgrade(ref self: TState, impl_hash: ClassHash);
+    fn upgrade_and_call(
+        ref self: TState, impl_hash: ClassHash, selector: felt252, calldata: Span<felt252>
+    );
+}
```

Comment:
I think it makes sense to remove this trait if we move the implementation of these methods to the `InternalImpl`, and maybe add an interface file with `IUpgradeable`, and we can add in the docs that developers are supposed to implement this interface exposing the internal methods defined in this module. Wdyt?

Reply:
Mmm I got you. Yeah, I initially had a similar approach, but I talked myself into creating a separate trait and impl in the contract (following our pattern of implementing interfaces). I'll update